cmd/cloud-endpoints-controller/main.go

```diff
@@ -364,7 +392,34 @@ func calcParentSig(parent *CloudEndpoint, addStr string) string {
 	return fmt.Sprintf("%x", hasher.Sum(nil))
 }
 
-func makeDefaultOpenAPISpec(endpoint, target string, jwtAudiences []string) (string, error) {
+func validateOpenAPISpec(specOriginal string) error {
+	var spec map[string]interface{}
+	return yaml.Unmarshal([]byte(specOriginal), &spec)
+}
+
+func getUserSuppliedOpenAPISpec(parent *CloudEndpoint, status *CloudEndpointControllerStatus) (string, error) {
+	var specString string
+	var err error
+	if parent.Spec.OpenAPISpec != nil {
+		specYaml, err := yaml.Marshal(parent.Spec.OpenAPISpec)
+		if err != nil {
+			log.Printf("[ERROR][%s] Supplied spec: %s", parent.Name, err.Error())
+			return "", errors.Wrap(err, "Could not load user supplied API from CloudEndpoint Spec")
+		}
+		specString = string(specYaml)
+	} else if name, key := parent.Spec.OpenAPISpecConfigMap.Name, parent.Spec.OpenAPISpecConfigMap.Key; name != "" && key != "" {
+		specString, err = getConfigMapSpecData(parent.ObjectMeta.Namespace, name, key)
+		if err != nil {
+			log.Printf("[ERROR][%s] error retrieving openapi spec from config map: %s, %s", parent.Name, name, err.Error())
+			err = errors.Wrap(err, "Could not get configMap for API Spec.")
+			return specString, err
+		}
+		status.ConfigMapHash = toSha1(specString)
```

Comment:
status.ConfigMapHash should be set at the controller level, not within this function. You don't need to pass status to this function.

Set the configmaphash after the spec has been validated in the main loop.